### PR TITLE
Refactoring rustc's driver once again

### LINKED ISSUE #35: Rework rustc driver
The rustc driver tested out several ways how an AST can be translated and how callbacks can be implemented. Now that some progress has been made on #8 we have an almost set API for how lint crates and drivers communicate.

The next step is now to restructure or rewrite the rustc driver to be maintainable and follow this new API.

Goals:
* [x] Implement a clear structure
* [ ] Document the structure
* [x] Improve maintainability
* [x] Add caching (Even if minimal)
* [x] Improve lifetime structure
	* This can still be improved but is already way better :)

### PR DESCRIPTION
**Pull request by xFrednet**
This PR refactored the entire conversion code of rustc's driver to use `ConversionContext` structs and documents the whole thing. I've renamed most functions and moved some of them. The functionality stayed the same. This feels like a big step in the right direction, but also took a good amount of time. FUN, it was a good distraction from studying ^^

@Niki4tap I'm uncertain if this is worth reviewing, as it's just countless surface changes, which are hard to review. If you want to help, I'd suggest checking the documentation updates and then have good faith in my refactoring abilities ^^. As always, a review is appreciated, but optional :)

Oh, last but not least:

Closes #35

### CODE REVIEW

**Review by Niki4tap — CHANGES_REQUESTED** [review]
> it was a good distraction from studying ^^

Distracts me as well :)

File: docs/internal/rustc_driver/conversion.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+An example is `SpanSourceInfo` required to create the API spans and convert them back to rustc's
+representation.
+
+Structs used by both conversion directions, like ID layouts, are stored inside the `common`
+submodule. The conversion context instances should be independent and in theory don't know
+of the other one.
+
+Conversion functions generally use the `to_` prefix, followed by the target object. `to_symbol_id`
+for example, takes a symbol and converts it to a `SymbolId` the direction is implied by the context
+that the function is defined in. For common conversion targets, the `_from_` prefix can be used to
```
[the remark this answers is not in the document]
Comment (xFrednet) [reply docs/internal/rustc_driver/conversion.md:27]: I meant to write suffix, even if that is not quite right. I can call it infix or just say that it gets extended with the source type after a `_from_` or somthing :)

**Comment by xFrednet** [comment]
> > it was a good distraction from studying ^^
> 
> Distracts me as well :)

You're welcome? ^^

All nits should have been fixed by the last commit. Thank you for checking!

**Comment by Niki4tap** [comment]
> You're welcome? ^^

Marker is more interesting than whatever I'm studying anyways.

> All nits should have been fixed by the last commit.

Thanks!

bors r+

**Comment by xFrednet** [comment]
I spent most of my lectures, half listening to the instructor, half working on marker. So, I know what you mean xD

**Review by Niki4tap — APPROVED** [review]
Apparently I need to approve now before merging :)

bors r+

**Comment by xFrednet** [comment]
I think you first have to mark the PR as approved to override the requested changes? I'll see if I can change that setting :)

**Comment by xFrednet** [comment]
Nope, I couldn't find a setting for it. Just reapproving it shouldn't be a problem :)

Thank you for the review!

**Pull request merged**